File: server/src/storage/delphi/keys.ts
/**
 * Key construction and shared constants for Delphi Storage V2.
 *
 * TypeScript twin of delphi/delphi_storage/keys.py — the rules here are part
 * of the cross-language conformance contract
 * (delphi/delphi_storage/conformance/README.md); keep the two in sync.
 */
import { InvalidError } from "./errors";
import type { RunManifest } from "./interface";

export const ENTITY_RUNS = "runs";
export const ENTITY_RUN_INPUTS = "run_inputs";
export const ENTITY_ARTIFACTS = "artifacts";
export const ENTITY_LATEST = "latest";
export const ENTITY_TOPIC_MODERATION = "topic_moderation";
export const ENTITY_COLLECTIVE_STATEMENTS = "collective_statements";

/**
 * Entities addressable through the generic put/get/query/delete operations.
 * `runs` and `latest` are mutated exclusively through the semantic ops so
 * their invariants (optimistic lock, monotonic seq) cannot be bypassed.
 */
export const GENERIC_ENTITIES = [
  ENTITY_RUN_INPUTS,
  ENTITY_ARTIFACTS,
  ENTITY_TOPIC_MODERATION,
  ENTITY_COLLECTIVE_STATEMENTS,
] as const;

export type GenericEntity = (typeof GENERIC_ENTITIES)[number];

/**
 * Reserved for backend-internal chunk rows (DynamoDB 400KB item limit);
 * forbidden in user-supplied keys.
 */
export const CHUNK_MARKER = "\u007f";

export const LOG_PREFIX = "log#";

const TS_RE = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z$/;

/** Validate the canonical timestamp format YYYY-MM-DDTHH:MM:SS.mmmZ. */
export function validateTs(ts: string): string {
  if (typeof ts !== "string" || !TS_RE.test(ts)) {
    throw new InvalidError(`timestamp must be YYYY-MM-DDTHH:MM:SS.mmmZ, got ${JSON.stringify(ts)}`);
  }
  if (Number.isNaN(Date.parse(ts))) {
    throw new InvalidError(`timestamp is not a real date: ${ts}`);
  }
  return ts;
}

export function nowTs(): string {
  // Date#toISOString always emits the canonical YYYY-MM-DDTHH:MM:SS.mmmZ form.
  return new Date().toISOString();
}

export function tsAddSeconds(ts: string, seconds: number): string {
  validateTs(ts);
  return new Date(Date.parse(ts) + seconds * 1000).toISOString();
}

/**
 * Compare strings by UTF-8 byte order (= Unicode code-point order, and
 * DynamoDB's native range-key order). NOT the same as JavaScript's default
 * `<` comparison, which uses UTF-16 code units and misorders astral-plane
 * characters relative to U+E000..U+FFFF.
 */
export function utf8Compare(a: string, b: string): number {
  return Buffer.compare(Buffer.from(a, "utf8"), Buffer.from(b, "utf8"));
}

/**
 * Claim-order string: ascending sort = highest priority first, then FIFO,
 * then job_id as tiebreak.
 */
export function claimOrder(priority: number, enqueuedAt: string, jobId: string): string {
  const p = Math.min(Math.max(Math.trunc(priority), 0), 9999);
  return `${String(9999 - p).padStart(4, "0")}#${enqueuedAt}#${jobId}`;
}

export function scopeForZid(zid: number, jobType: string): string {
  return `zid#${zid}#${jobType}`;
}

export function scopeForRid(rid: number, jobType: string): string {
  return `rid#${rid}#${jobType}`;
}

/**
 * Latest-pointer scopes a completed run flips (design §4.2 entity 4).
 * IMPORTED runs never auto-flip: the backfill importer advances pointers
 * explicitly under the no-clobber invariant (design §6.2 invariant 1).
 */
export function scopesForRun(run: RunManifest): string[] {
  if (run.job_type === "FULL_PIPELINE") {
    if (run.zid === null || run.zid === undefined) {
      throw new InvalidError(`run ${run.job_id}: FULL_PIPELINE requires zid`);
    }
    return [scopeForZid(run.zid, run.job_type)];
  }
  if (run.job_type === "NARRATIVE_BATCH" || run.job_type === "SERVER_NARRATIVE") {
    if (run.rid === null || run.rid === undefined) {
      throw new InvalidError(`run ${run.job_id}: ${run.job_type} requires rid`);
    }
    return [scopeForRid(run.rid, run.job_type)];
  }
  if (run.job_type === "IMPORTED") {
    return [];
  }
  throw new InvalidError(`run ${run.job_id}: unknown job_type ${JSON.stringify(run.job_type)}`);
}

export function logSk(seq: number): string {
  return `${LOG_PREFIX}${String(seq).padStart(8, "0")}`;
}

/**
 * Compose an artifact sort key from parts using the existing '#' composite
 * convention (e.g. artifactKey('umap', 'topic', 0, 3)).
 */
export function artifactKey(...parts: Array<string | number>): string {
  const strs = parts.map((p) => String(p));
  for (const s of strs) {
    if (!s) throw new InvalidError("artifact key parts must be non-empty");
    if (s.includes(CHUNK_MARKER)) {
      throw new InvalidError("artifact key parts must not contain U+007F");
    }
  }
  return strs.join("#");
}

File: delphi/delphi_storage/keys.py
"""Key construction and shared constants for Delphi Storage V2.

The rules here are part of the cross-language conformance contract
(delphi_storage/conformance/README.md); server/src/storage/delphi/keys.ts
mirrors them exactly.
"""

import re
from datetime import datetime, timedelta, timezone
from typing import TYPE_CHECKING, Union

if TYPE_CHECKING:  # pragma: no cover - typing only, avoids a runtime cycle
    from delphi_storage.models import JobType, RunManifest

ENTITY_RUNS = "runs"
ENTITY_RUN_INPUTS = "run_inputs"
ENTITY_ARTIFACTS = "artifacts"
ENTITY_LATEST = "latest"
ENTITY_TOPIC_MODERATION = "topic_moderation"
ENTITY_COLLECTIVE_STATEMENTS = "collective_statements"

# Entities addressable through the generic put/get/query/delete operations.
# `runs` and `latest` are mutated exclusively through the semantic ops so
# their invariants (optimistic lock, monotonic seq) cannot be bypassed.
GENERIC_ENTITIES = (
    ENTITY_RUN_INPUTS,
    ENTITY_ARTIFACTS,
    ENTITY_TOPIC_MODERATION,
    ENTITY_COLLECTIVE_STATEMENTS,
)

ALL_ENTITIES = (ENTITY_RUNS, ENTITY_LATEST) + GENERIC_ENTITIES

# Reserved for backend-internal chunk rows (DynamoDB 400KB item limit);
# forbidden in user-supplied keys.
CHUNK_MARKER = "\x7f"

KEY_SEPARATOR = "#"

LOG_PREFIX = "log#"

_TS_RE = re.compile(r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z$")
_TS_FORMAT = "%Y-%m-%dT%H:%M:%S.%fZ"


def validate_ts(ts: str) -> str:
    """Validate the canonical timestamp format YYYY-MM-DDTHH:MM:SS.mmmZ."""
    if not isinstance(ts, str) or not _TS_RE.match(ts):
        raise ValueError(f"timestamp must be YYYY-MM-DDTHH:MM:SS.mmmZ, got {ts!r}")
    datetime.strptime(ts, _TS_FORMAT)  # reject e.g. month 13
    return ts


def format_ts(dt: datetime) -> str:
    dt = dt.astimezone(timezone.utc)
    return dt.strftime("%Y-%m-%dT%H:%M:%S.") + f"{dt.microsecond // 1000:03d}Z"


def now_ts() -> str:
    return format_ts(datetime.now(timezone.utc))


def ts_add_seconds(ts: str, seconds: float) -> str:
    validate_ts(ts)
    dt = datetime.strptime(ts, _TS_FORMAT).replace(tzinfo=timezone.utc)
    return format_ts(dt + timedelta(seconds=seconds))


def _job_type_value(job_type: Union[str, "JobType"]) -> str:
    return getattr(job_type, "value", job_type)


def claim_order(priority: int, enqueued_at: str, job_id: str) -> str:
    """Claim-order string: ascending sort = highest priority first, then FIFO,
    then job_id as tiebreak."""
    p = min(max(int(priority), 0), 9999)
    return f"{9999 - p:04d}#{enqueued_at}#{job_id}"


def scope_for_zid(zid: int, job_type: Union[str, "JobType"]) -> str:
    return f"zid#{zid}#{_job_type_value(job_type)}"


def scope_for_rid(rid: int, job_type: Union[str, "JobType"]) -> str:
    return f"rid#{rid}#{_job_type_value(job_type)}"


def scopes_for_run(run: "RunManifest") -> list[str]:
    """Latest-pointer scopes a completed run flips (design §4.2 entity 4).

    IMPORTED runs never auto-flip: the backfill importer advances pointers
    explicitly under the no-clobber invariant (design §6.2 invariant 1).
    """
    job_type = _job_type_value(run.job_type)
    if job_type == "FULL_PIPELINE":
        if run.zid is None:
            raise ValueError(f"run {run.job_id}: FULL_PIPELINE requires zid")
        return [scope_for_zid(run.zid, job_type)]
    if job_type in ("NARRATIVE_BATCH", "SERVER_NARRATIVE"):
        if run.rid is None:
            raise ValueError(f"run {run.job_id}: {job_type} requires rid")
        return [scope_for_rid(run.rid, job_type)]
    if job_type == "IMPORTED":
        return []
    raise ValueError(f"run {run.job_id}: unknown job_type {job_type!r}")


def log_sk(seq: int) -> str:
    return f"{LOG_PREFIX}{seq:08d}"


def artifact_key(*parts: Union[str, int]) -> str:
    """Compose an artifact sort key from parts using the existing '#'
    composite convention (e.g. artifact_key('umap', 'topic', 0, 3))."""
    strs = [str(p) for p in parts]
    for s in strs:
        if not s:
            raise ValueError("artifact key parts must be non-empty")
        if CHUNK_MARKER in s:
            raise ValueError("artifact key parts must not contain U+007F")
    return KEY_SEPARATOR.join(strs)
